smartsim/_core/utils/security.py

```diff
@@ -0,0 +1,289 @@
+        the public and private key pairs for servers & clients"""
+        for locator in [self._server_locator, self._client_locator]:
+            if not locator.public_dir.exists():
+                locator.public_dir.mkdir(parents=True, mode=_KeyPermissions.WORLD_R)
+
+            if not locator.private_dir.exists():
+                locator.private_dir.mkdir(parents=True, mode=_KeyPermissions.OWNER_FULL)
+
+    @classmethod
+    def _load_keypair(cls, locator: _KeyLocator, in_context: bool) -> KeyPair:
+        """Load a specific `KeyPair` from disk
+
+        :param locator: a `KeyLocator` that specifies the path to an existing key
+        :type locator: KeyLocator
+        :param in_context: Boolean flag indicating if the keypair is the active
+        context; ensures the public and private keys are both loaded when `True`.
+        Only the public key is loaded when `False`
+        :type in_context: bool
+        :returns: a KeyPair containing the loaded public/private key
+        :rtype: KeyPair
+        """
+        # private keys contain public & private key parts
+        key_path = locator.private if in_context else locator.public
+        pub_key, priv_key = zmq.auth.load_certificate(key_path)
+
+        # avoid a `None` value in the private key when it isn't loaded
+        return KeyPair(pub_key, priv_key or b"")
+
+    def _load_keys(self) -> t.Tuple[KeyPair, KeyPair]:
+        """Use ZMQ auth to load public/private key pairs for the server and client
+        components from the standard key paths for the associated experiment
+
+        :returns: 2-tuple of `KeyPair` (server_keypair, client_keypair)
+        :rtype: Tuple[KeyPair, KeyPair]"""
+        try:
+            server_keys = self._load_keypair(self._server_locator, self._as_server)
+            client_keys = self._load_keypair(self._client_locator, self._as_client)
+
+            return server_keys, client_keys
+        except (ValueError, OSError):
+            # expected if no keys could be loaded from disk
+            ...
+
+        return KeyPair(), KeyPair()
+
+    @classmethod
+    def _move_public_key(cls, locator: _KeyLocator) -> None:
+        """The public and private key pair are created in the same directory. Move
+        the public key out of the private subdir and into the public subdir
+
+        :param locator: `KeyLocator` that determines the path to the
+        key pair persisted in the same directory.
+        :type locator: KeyLocator"""
+        new_path = locator.private.with_suffix(locator.public.suffix)
+        if new_path != locator.public:
+            new_path.rename(locator.public)
+
+    def _create_keys(self) -> None:
+        """Create and persist key files to disk"""
+        for locator in [self._server_locator, self._client_locator]:
+            # create keys in the private directory...
+            zmq.auth.create_certificates(locator.private.parent, locator.private.stem)
```

Comment:
Since we have the attrs, can we use
```suggestion
            zmq.auth.create_certificates(locator.private_dir, locator.private.stem)
```
for just a hair more readablility?

Reply:
Great eye. Swapped to use property instead.